### PR TITLE
Refactor proxy state tests

### LINKED ISSUE #39894: [serve] refactor `test_proxy_state.py` to use actor proxy wrapper
Most of the tests in `test_proxy_state.py` can be refactored not to rely on ray. Refactor them to make better tests and run faster.

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Refactor `test_proxy_state` tests to use the new proxy wrapper interface. Also fixed a small bug where when the health check gets `ProxyWrapperCallStatus.FINISHED_FAILED`, it skipped 3 consecutive failure retries.  

This test now runs in just under 3s locally 😄
image

## Related issue number

Closes https://github.com/ray-project/ray/issues/39894

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/test_proxy_state.py [hunk 1]
```diff
@@ -41,84 +39,92 @@ def get_alive_node_ids(self):
...
 
-@pytest.fixture
-def all_nodes(number_of_worker_nodes) -> List[Tuple[str, str]]:
-    return [(HEAD_NODE_ID, "fake-head-ip")] + [
-        (f"worker-node-id-{i}", f"fake-worker-ip-{i}")
-        for i in range(number_of_worker_nodes)
-    ]
+class FakeProxyWrapper(ProxyWrapper):
+    def __init__(self, actor_handle=None, *args, **kwargs):
+        self.actor_handle = actor_handle or FakeProxyActor(*args, **kwargs)
```
Comment (edoakes) [inline python/ray/serve/tests/test_proxy_state.py:55]: why do we need to support a real actor_handle here? is this temporary as you continue refactoring tests?

Comment (GeneDer) [reply python/ray/serve/tests/test_proxy_state.py:49]: oops, it was for `test_proxy_state_update_healthy_check_health_sometimes_fails`. Testing the behavior where the health check is called x number of times in between proxy failing and succeeding health checks

File: python/ray/serve/tests/test_proxy_state.py [hunk 2]
```diff
@@ -41,84 +34,96 @@ def get_alive_node_ids(self):
...
+    cluster_node_info_cache=MockClusterNodeInfoCache(),
+    actor_proxy_wrapper_class=FakeProxyWrapper,
+) -> (ProxyStateManager, ClusterNodeInfoCache):
+    return (
+        ProxyStateManager(
+            SERVE_CONTROLLER_NAME,
+            config=http_options,
+            head_node_id=head_node_id,
+            cluster_node_info_cache=cluster_node_info_cache,
+            proxy_actor_class=FakeProxyActor,
```
Comment (edoakes) [inline python/ray/serve/tests/test_proxy_state.py:112]: looks like this can be totally removed now :)

Comment (GeneDer) [reply python/ray/serve/tests/test_proxy_state.py:112]: oh that's right, let me take it out, too 😄

**Review by edoakes — APPROVED** [review]

**Pull request merged**